prov/psm2/src/psmx2.h

```diff
@@ -562,6 +566,32 @@ struct psmx2_fid_domain {
 	uint32_t		max_atomic_size;
 
 	struct dlist_entry	entry;
+
+	/* Lock/Unlock function pointers set based on FI_THREAD model */
+	psmx2_lock_fn_t av_lock_fn;
+	psmx2_lock_fn_t am_req_pool_lock_fn;
+	psmx2_lock_fn_t trx_ctxt_lock_fn;
+	psmx2_lock_fn_t rma_queue_lock_fn;
+	psmx2_lock_fn_t trigger_queue_lock_fn;
+	psmx2_lock_fn_t peer_lock_fn;
+	psmx2_lock_fn_t sep_lock_fn;
+	psmx2_lock_fn_t trigger_lock_fn;
+	psmx2_lock_fn_t cq_lock_fn;
+	psmx2_lock_fn_t mr_lock_fn;
+	psmx2_lock_fn_t context_lock_fn;
+	psmx2_trylock_fn_t poll_trylock_fn;
+	psmx2_unlock_fn_t av_unlock_fn;
+	psmx2_unlock_fn_t am_req_pool_unlock_fn;
+	psmx2_unlock_fn_t trx_ctxt_unlock_fn;
+	psmx2_unlock_fn_t rma_queue_unlock_fn;
+	psmx2_unlock_fn_t trigger_queue_unlock_fn;
+	psmx2_unlock_fn_t peer_unlock_fn;
+	psmx2_unlock_fn_t sep_unlock_fn;
+	psmx2_unlock_fn_t trigger_unlock_fn;
+	psmx2_unlock_fn_t cq_unlock_fn;
+	psmx2_unlock_fn_t mr_unlock_fn;
+	psmx2_unlock_fn_t context_unlock_fn;
+	psmx2_unlock_fn_t poll_unlock_fn;
```

Comment:
This is a massive number of different locks. Acquiring and releasing this many locks in a fast path would have a huge performance hit.

Can we remove any locks that aren't considered part of the fast path?  Also, try to reduce the number of locks greatly, and hold them for longer if needed.

Reply:
It's 12 locks in total, not 24 locks. Some of them are not in the fast path, like the lock for the peer list and the lock for the sep list. For those in the fast path, many of them are used under different scenarios. For example,  lock for am_req_pool is only used for RMA operations; lock for rma_queue is only used for large size RMA operations; lock for triggered queue is only used for triggered ops; etc.

Reply:
Btw, I don't object to the changes.  Since this is increasing the size of the domain, I don't see that as a big deal.  It's the number of locks that worries me.  Some of the API calls are always intended to be thread safe.

Reply:
okay I can't count locks

Reply:
I would move the lock/unlock pointers for the same lock together, so that the pointers hit the same cache line.  Otherwise, changes are fine.

Reply:
> I would move the lock/unlock pointers for the same lock together, so that the pointers hit the same cache line. Otherwise, changes are fine.

So, group the pointers to be listed as:
psmx2_lock_fn_t av_lock_fn;
psmx2_unlock_fn_t av_unlock_fn;
etc.

correct?

Reply:
correct

Reply:
> correct

Done, thanks!